File: tools/grammar/driver.py
# The Python half of the byte-identical gate: the same walk as
# driver-body.raku, through the rakulang package over librakupp. Run by
# grammar-smoke.raku; outputs are compared byte for byte.
#
#   python3 driver.py <librakupp-path> <grammar-file> <input-file>

import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)),
                                "..", "..", "bindings", "python"))
import rakulang

lib, grammar_file, input_file = sys.argv[1], sys.argv[2], sys.argv[3]


def canon(x):
    if isinstance(x, list):
        return "[" + ",".join(canon(v) for v in x) + "]"
    if isinstance(x, dict):
        return "{" + ",".join(k + ":" + canon(x[k]) for k in sorted(x)) + "}"
    return str(x)


def raku_bool(b):
    return "True" if b else "False"


it = rakulang.interpreter(lib)
g = rakulang.Grammar.from_file(grammar_file, name="Log", actions="LogActions")

with open(input_file, encoding="utf-8") as f:
    m = g.parse(f.read())
assert m, "gate: the log corpus did not parse"

lines = m["line"]
print("lines", len(lines))
print("islist", raku_bool(lines._walk("islist")))

for line in lines:
    print(line["ip"].str(), line["status"].str())

print("made", m["line"][0]["size"].made)
print("req.str", m["line"][42]["req"].str())
print("size.int", m["line"][42]["size"].int())
print("missing", raku_bool(bool(m["nope"])))
print("tree", canon(m["line"][999].tree()))

one = g.parse('7.7.7.7 - - [x] "GET / HTTP/1.1" 200 5\n', rule="line")
print("rule-parse", one["status"].str())

bad = g.parse("this is not a log line")
print("failed-parse", "Match" if bad else "None")

try:
    g.parse("this is not a log line", strict=True)
    print("diag none")
except rakulang.ParseError as e:
    print(f"diag line {e.line} col {e.column} rule {e.rule}")

File: tools/grammar/driver.mjs
// The JS half of the byte-identical gate: the same walk as driver.py,
// through bindings/js/rakulang.js on bun:ffi. Run by grammar-smoke.raku;
// outputs are compared byte for byte.
//
//   bun tools/grammar/driver.mjs <librakupp-path> <grammar-file> <input-file>

import { Grammar, interpreter, ParseError } from "../../bindings/js/rakulang.js";
import { readFileSync } from "fs";

const [lib, grammarFile, inputFile] = process.argv.slice(2);

function canon(x) {
  if (Array.isArray(x)) return "[" + x.map(canon).join(",") + "]";
  if (x !== null && typeof x === "object")
    return "{" + Object.keys(x).sort().map(k => k + ":" + canon(x[k])).join(",") + "}";
  return String(x);
}

const rakuBool = b => (b ? "True" : "False");

interpreter(lib);
const g = Grammar.fromFile(grammarFile, { name: "Log", actions: "LogActions" });

const m = g.parse(readFileSync(inputFile, "utf8"));
if (!m) { console.error("gate: the log corpus did not parse"); process.exit(1); }

const lines = m.get("line");
let out = [];
out.push(`lines ${lines.length}`);
out.push(`islist ${rakuBool(lines.isList())}`);

for (const line of lines) out.push(`${line.get("ip").str()} ${line.get("status").str()}`);

out.push(`made ${m.get("line").at(0).get("size").made()}`);
out.push(`req.str ${m.get("line").at(42).get("req").str()}`);
out.push(`size.int ${m.get("line").at(42).get("size").int()}`);
out.push(`missing ${rakuBool(m.get("nope").truthy())}`);
out.push(`tree ${canon(m.get("line").at(999).tree())}`);

const one = g.parse('7.7.7.7 - - [x] "GET / HTTP/1.1" 200 5\n', { rule: "line" });
out.push(`rule-parse ${one.get("status").str()}`);

const bad = g.parse("this is not a log line");
out.push(`failed-parse ${bad ? "Match" : "None"}`);

try {
  g.parse("this is not a log line", { strict: true });
  out.push("diag none");
} catch (e) {
  if (!(e instanceof ParseError)) throw e;
  out.push(`diag line ${e.line} col ${e.column} rule ${e.rule}`);
}

console.log(out.join("\n"));
